### ISSUE TITLE
Consider adding Method.Supports(Function) (bool, error)

### ISSUE THREAD

**Issue by vladimir-ch** [gonum/optimize#48]
@btracey 's comment from #45 : Given we do use some magic, I can see the need for support. It seems like the better case is for the optimizer to have a Support function that either returns a boolean (if it supports), or an error which is nil if it supports, and provides a detailed explanation if it isn't, i.e. "Bfgs requires the Function to have a Gradient. Function does not implement either Gradient or FunctionGradient."

**Comment by vladimir-ch** [comment]
A related point is whether some decisions should be based on what the objective function supports or on what the optimizer requires. For example, currently some decisions are based on whether `loc.Gradient` is `nil` or not, or on values in FunctionInfo (which is the same in the end). Examples of such are whether GradientEvaluation should be requested before MajorIteration in Linesearch, or whether gradient norm should be printed in Printer. However, if a gradient-free method is used, then loc.Gradient should not be updated even if the function supports, or should it?

**Comment by btracey** [comment]
No, it shouldn’t, which is why we have both checks. I’d have to re-read the code to be sure, but I think we’ve been pretty careful about correctly handling the case where you’re using a non-gradient based optimizer on a gradient-based function.

**Comment by vladimir-ch** [comment]
Well, at the moment the only optimizer we have is Linesearch which requests a gradient evaluation on a subiteration whenever loc.Gradient is not nil (no matter if NextDir or LinesearchMethod need it). And it is not nil whenever the function can evaluate the gradient. So only the function decides if the gradient is evaluated.

**Comment by vladimir-ch** [comment]
As a next step in my effort to Hessianize the code I would like to clear up this issue. To sum it up, the problem with the current code is that we allocate Location based on capabilities of given Function, not on what Method actually needs. For example, when we pass a function with a gradient to a gradient-free method, not only the gradient is allocated and then invalidated with NaNs at each evaluation, there are also several decisions based on whether loc.Gradient is nil or not (e.g., if and what kind of EvaluationType we request at SubIteration). So I hesitate to finally add Hessian to Location knowing that I won't be able to reliably tell whether to allocate it or not.

The only solution I see is for Method be able to advertise its requirements by adding a new interface method to it. And `Method.Supports(Function) (bool, error)` does not seem right to me, because gradient-free method will support a function with a Hessian, but that does not solve this issue. I have been thinking about this recently and as the best minimal solution I see adding something like:
```
Requires() (grad, hess bool)
```
to Method. `Linesearch` struct would no longer be a `Method` but that is not a problem, it would still be a useful helper struct as it is now.

Alternatively, Requires() could return FunctionInfo, but that returns too much (we don't care about IsStatuser) and is too ambiguous (when IsGradient is false, but IsFunctionGradient is true).

Any comments or alternative approaches/solutions, @kortschak @btracey ?

**Comment by kortschak** [comment]
I think `Requires() (grad, hess bool)` looks like a fairly nice simple solution.

**Comment by btracey** [comment]
I like the idea, but I think we should make sure to consider the solution in the context of constrained optimization. There are two factors; what the optimizer requires and what the optimizer supports. Trying to use an optimizer on a problem it doesn't support is an error, and Local evaluating more than it needs is a bug (of sorts). I think the list of method options are
````
Function
Gradient
Hessian
Bounded Constraints
Linear Equality/Inequality
Nonlinear Equality/Inequality
````
The interface we design should, for example, allow BFGS to communicate that it needs the function value and the gradient, but cannot support a constrained function of any kind. It seems like a reasonable API would be along the lines of:
````
Requires() (grad, hess, bool)
Supports() (bounds, linEq, linIneq, nonlinEq, nonlinIneq bool) // Probably make this a struct
````
BFGS could then return false for all of the Support returns. (LBFGS-B, for example, would support bounded). The catch is that some optimizers require variable bounds to work properly (some GAs for example). This suggests adding "bounds" to Requires().

There may be a better way to combine these ideas, which is why I bring it up.

**Comment by vladimir-ch** [comment]
I can see that `Supports() (...)` could be potentially useful, but I would prefer to wait until we actually have constrained optimization (the same for adding "bounds" to Requires()). 

For the moment I will add `Requires() (grad, hess bool)` to `Method` and see how it works.
